[CI] ConcurrentSeqNoVersioningIT testSeqNoCASLinearizability failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/hdrkqs5hmpp6k/tests/:server:internalClusterTest/org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT/testSeqNoCASLinearizability

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT.testSeqNoCASLinearizability" -Dtests.seed=FF0D42993DBCEB07 -Dtests.locale=es-HN -Dtests.timezone=Chile/Continental -Druntime.java=18`

**Applicable branches:**
8.3

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.versioning.ConcurrentSeqNoVersioningIT&tests.test=testSeqNoCASLinearizability

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=4099, name=elasticsearch[node_t3][generic][T#3], state=RUNNABLE, group=TGRP-ConcurrentSeqNoVersioningIT]


  Caused by: java.lang.AssertionError: supposedly in-sync shard copy received a global checkpoint [566] that is higher than its local checkpoint [560]

    at __randomizedtesting.SeedInfo.seed([FF0D42993DBCEB07]:0)
    at org.elasticsearch.index.shard.IndexShard.updateGlobalCheckpointOnReplica(IndexShard.java:2760)
    at org.elasticsearch.index.shard.IndexShard.lambda$innerAcquireReplicaOperationPermit$30(IndexShard.java:3537)
    at org.elasticsearch.index.shard.IndexShard$5.onResponse(IndexShard.java:3398)
    at org.elasticsearch.index.shard.IndexShard$5.onResponse(IndexShard.java:3368)
    at org.elasticsearch.index.shard.IndexShard.lambda$asyncBlockOperations$24(IndexShard.java:3319)
    at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
    at org.elasticsearch.index.shard.IndexShardOperationPermits$1.doRun(IndexShardOperationPermits.java:107)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:769)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:833)

```

--- a/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
@@ -517,6 +517,7 @@ public final class MockTransportService extends TransportService {
                     RequestHandlerRegistry<?> reg = MockTransportService.this.getRequestHandler(action);
                     clonedRequest = reg.newRequest(bStream.bytes().streamInput());
                 }
+                assert clonedRequest.getClass().equals(request.getClass()) : clonedRequest + " vs " + request;
 
                 final RunOnce runnable = new RunOnce(new AbstractRunnable() {
                     @Override

--- a/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
@@ -510,6 +510,8 @@ public final class MockTransportService extends TransportService {
                     // Some request handlers read back a BytesTransportRequest
                     // into a different class that cannot be re-serialized (i.e. JOIN_VALIDATE_ACTION_NAME),
                     // in those cases we just copy the raw bytes back to a BytesTransportRequest.
+                    // This is only needed for the BwC for JOIN_VALIDATE_ACTION_NAME and can be removed in the next major
+                    assert Version.CURRENT.major == Version.V_7_17_0.major + 1;
                     clonedRequest = new BytesTransportRequest(bStream.bytes().streamInput());
                 } else {
                     RequestHandlerRegistry<?> reg = MockTransportService.this.getRequestHandler(action);

--- a/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
@@ -37,6 +37,7 @@ import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.test.tasks.MockTaskManager;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.tracing.Tracer;
+import org.elasticsearch.transport.BytesTransportRequest;
 import org.elasticsearch.transport.ClusterConnectionManager;
 import org.elasticsearch.transport.ConnectTransportException;
 import org.elasticsearch.transport.ConnectionProfile;
@@ -502,10 +503,18 @@ public final class MockTransportService extends TransportService {
                 }
 
                 // poor mans request cloning...
-                RequestHandlerRegistry<?> reg = MockTransportService.this.getRequestHandler(action);
                 BytesStreamOutput bStream = new BytesStreamOutput();
                 request.writeTo(bStream);
-                final TransportRequest clonedRequest = reg.newRequest(bStream.bytes().streamInput());
+                final TransportRequest clonedRequest;
+                if (request instanceof BytesTransportRequest) {
+                    // Some request handlers read back a BytesTransportRequest
+                    // into a different class that cannot be re-serialized (i.e. JOIN_VALIDATE_ACTION_NAME),
+                    // in those cases we just copy the raw bytes back to a BytesTransportRequest.
+                    clonedRequest = new BytesTransportRequest(bStream.bytes().streamInput());
+                } else {
+                    RequestHandlerRegistry<?> reg = MockTransportService.this.getRequestHandler(action);
+                    clonedRequest = reg.newRequest(bStream.bytes().streamInput());
+                }
 
                 final RunOnce runnable = new RunOnce(new AbstractRunnable() {
                     @Override

